File: Algorithms/Fibonacci Search/fibonacci_search.py
###################################################
### Author : Divyesh Mehta
### Algorithm : Fibonacci Search
### Language : Python
####################################################

def fibonacci_search(arr, e, n):
    n2 = 0
    n1 = 1
    nT = n2 + n1

    while nT < n:
        n2 = n1
        n1 = nT
        nT = n2 + n1

    flag = -1

    while nT > 1:
        i = min(flag + n2, n - 1)

        if arr[i] < e:
            nT = n1
            n1 = n2
            n2 = nT - n1
            flag = i

        elif arr[i] > e:
            nT = n2
            n1 = n1 - n2
            n2 = nT - n1
        else:
            print(f"Element found at index {i}")
            return

    if n1 and arr[flag + 1] == e:
        return flag + 1

    print("Given Input Not Found in Array !!")


arr = [10, 20, 30, 40, 50]

num = int(input("Enter The Value to Search:"))

fibonacci_search(arr, num, len(arr))

File: Algorithms/Fibonacci Search/Fibonacci_Search.cpp
#include <iostream> 
using namespace std;

int fibonacciSearch(int arr[], int x, int n){
    int m2 = 0;
    int m1 = 1;
    int m = m2+m1;

    while (n > m){
        m2 = m1;
        m1 = m;
        m = m1+m2;
    }

    int offset = -1;

    while (m>1){
        int i = min(offset+m2, n-1);

        if (arr[i] < x){
            m = m1;
            m1 = m2;
            m2 = m - m1;
            offset = i;
        }

        else if(arr[i] > x){
            m = m2;
            m1 = m1 - m;
            m2 = m - m1;
        }

        else{
            return i;
        }
    }

    if(arr[offset+1] == x){
        return offset+1;
    }

    return -1;
}

int main(){
    int arr[] = {-65, 1, 5, 9, 11, 54, 64, 85};
    int x = 85;
    int len = 8;

    int result = fibonacciSearch(arr, x, len);
    if (result==-1){
        cout<<"Element not found."<<endl;
    } else{
        cout<<"Element found at index "<<result<<endl;
    }

    return 0;
}
